Binomial self report

jeremy-haynes opened the pull request:
Hey Nate,

I started on the binomial model for the self-report data and am a little stuck, so I figured that a (hopefully) good way to see if you could look at it would be to make a new branch off of the [fix-group-means-correlations](https://github.com/hollysully/IGT_PP_Shared/pull/10#issue-1729588904) branch (let me know if there was a better way of doing this). Where I'm stuck is that the model produces a lot (~90%) divergent transitions (not good 🥲). I _think_ that it may have to do with the priors. I based them off of Kruschke, but I've also run into a similar issue before with trying to fit somewhat similar models, so I'm thinking I'm misunderstanding something.

I tried to setup the [RMD file](1_IGT_PP/Code/R/1_Analyses/binomial_selfreport.Rmd) so you could fit the [binomial stan model](1_IGT_PP/Code/Stan/binomial_selfreport.stan) for just the _prdep_tot_ data (the code is designed to run through each measure but there's a line that sets the scale of interest to just _prdep_tot_). There are some things I need to check with Tom about in terms of data, but I don't think those would affect the model-fit. Essentially, I've subtracted each scales' minimum possible score from the participant's score and the maximum possible score and I need him to confirm some of those mins/maxes (letting you know in case that does have some effect I'm unaware of).

Let me know what you think and if it'd be easier to meet about it I'd be happy to set up a time.

Thanks!
Jeremy

Nathaniel-Haines commented on 1_IGT_PP/Code/R/0_Preprocess/SelfReportPreprocess.R:
```diff
@@ -0,0 +1,169 @@
+                    min_item_score = c(0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1),
+                    max_item_score = c(4, 5, 5, 5, NA_real_, 4, 4, 4, 4, 4, 5, 5, 4),
+                    n_sessions = c(1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2)) %>% 
+  mutate(min = n_items * min_item_score,
+         max = n_items * max_item_score,
+         scaled_max = max - min)
+
+# Make scaled scores and check out which scores are below the minimums - for those scores, we'll truncate them to 0
+full_data = both_sessions %>% 
+  left_join(scales) %>% 
+  mutate(scaled_score = case_when(score-min < 0 ~ 0, is.numeric(score) ~ score - min))
```
I think scaling makes sense! For the binomial model, we would just want to make sure that each response increment is accounted for. e.g., answering the lowest on all items should give a 0, and answering the highest on all items should give a number that indicates each possible increment above 0. Basically, the first category for each item should be scored as 0, and every ordinal category above the first category should be scored with an additional +1 (let me know if I am explaining that sensibly ha)

jeremy-haynes replied:
That makes sense!... I'm pretty sure haha. Essentially, the participants' scores should be bounded between 0% (answering the lowest on all items) and 100% (answering the highest on all items) where 100% is entered in the model as the actual maximum score for that scale. The code above should make those adjustments. I still have to double-check some of those. Also, some participants had missing items, which put their score below 0% in some cases. For now, I truncated those to 0 which assumes they scored 0 on the missing item(s) which seemed reasonable to me (they scored '0' on all the other items). Let me know if you have other thoughts. I'm also going to bring this up with Tom today. Thanks!

Nathaniel-Haines reviewed (changes requested):
@jeremy-haynes this looks good! I added some notes on what needs to change to fix divergences—let me know if you have any questions on that.

Also, I merged-to-main the branch you have this one set to merge into.. 😅 That said, you should just be able to set this branch to merge to main as well!

Nathaniel-Haines commented:
> Also, I merged-to-main the branch you have this one set to merge into.. 😅 That said, you should just be able to set this branch to merge to main as well!

@jeremy-haynes I went ahead and changed it FYI

jeremy-haynes commented on 1_IGT_PP/Code/Stan/binomial_selfreport.stan:
```diff
@@ -0,0 +1,52 @@
+}
+
+
+parameters {
+  real kappaMinusTwo;
+  real omega;
+  vector<lower=0, upper=1>[N] theta; // specify theta as a vector with N values
+}
+
+model {
+  real kappa; // Declare transformed kappa hyperprior
+  
+  // PRIORS
+  // Hyperpriors
+  kappaMinusTwo ~ gamma(.01,.01);
+  kappa = kappaMinusTwo+2;
+  omega ~ beta(1,1);
+  
+  // Lower-level priors
+  theta ~ beta(omega*(kappa-2)+1, (1-omega)*(kappa-2)+1);
```
Ahh ok cool, this all makes sense thank you! I sometimes add code that I know is unnecessary to keep track of things but I hadn't thought of removing the `+1`'s. Does adding those restrict us from seeing a bimodal distribution of thetas if the data were bimodally distributed or would it just be more 'difficult' for the likelihood to detect that? I'm still trying to get a handle on non-normal priors (particularly hierarchical ones which I struggle with deeply). Anyway, that does make sense about adding the constraints: at least in the sense that I'd added them to the thetas.

jeremy-haynes replied:
Oh, and another question related to the gamma hyperprior (and hyperpriors more broadly). I struggle with wrapping my head around the hierarchical effect of the hyperprior on the lower-level prior(s). Do you have a suggestion for working through that? For example, when thinking about how the gamma prior on omega specifies the concentration of the beta priors on the thetas, I'm not entirely sure how to simulate the effect of that gamma prior on omega. One thing I have tried is to get the mean of a set of hyperprior(s) I'm considering and see their effect on the lower-level prior. Is that a good way of going about that?

jeremy-haynes reviewed (commented):
I think we took care of these

jeremy-haynes commented:
Hey @Nathaniel-Haines ,

I think I've got the ind-mu models for the [IGT](1_IGT_PP/Code/R/1_Analyses/fitting_ORL_indMu.Rmd) and [self-report](1_IGT_PP/Code/R/1_Analyses/binomial_selfReport_indMu.Rmd) data working. That is, the models where we use all of the data from both sessions in estimating each parameter individually (i.e., not a separate parameter for each session). So far, these models are fit to the IGT and self-report data separately (i.e., not as a joint model). I believe those are working correctly - at least there doesn't seem to be 90% divergent transitions (or other problems) after you helped on the last part haha.

I was feeling a bit confident, so I took a crack at the joint model where we wanted to fit the ORL to the IGT data and the binomial model to the self-report data simultaneously, so that we could obtain correlations between the ORL and binomial model parameters ([RMD here](1_IGT_PP/Code/R/1_Analyses/fitting_joint_ORL_binomial_indMu.Rmd) and [stan here](1_IGT_PP/Code/Stan/igt_selfreport_orl_binomial_joint_indMu.stan). I get an error that I cannot seem to diagnose (below). Because it's quite possible I am totally off the mark with this joint model, I figured now would be a good time to touch base with you so I don't waste time diagnosing this model if it's totally not what we were thinking). I'm happy to work through this via messaging or meeting - I could meet this week or next, whenever is convenient for you.

I also changed several file names in my [last commit](https://github.com/hollysully/IGT_PP_Shared/commit/901f2fdd5f6c5690463f16326221aa7cca3a2449). I did this because I was getting confused on what models the files were referencing (e.g., the linear model was called a joint ind-mu model etc.); however, that might mess with how some files are referenced for your gitignored files (e.g., the .rds models). Feel free to undo the last commit if that messes things up on your end too much.

Hope all is well!
Jeremy

Chain 1: Rejecting initial value:
Chain 1:   Log probability evaluates to log(0), i.e. negative infinity.
Chain 1:   Stan can't start sampling from this initial value.
...
Chain 1: Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:Chain 1: Exception: modela06c64065160__namespace::log_prob: theta[sym1__] is nan, but must be greater than or equal to 0 (in 'string', line 78, column 2 to column 36)
Chain 1: If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
Chain 1: but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.

Nathaniel-Haines commented on 1_IGT_PP/Code/R/1_Analyses/fitting_ORL_lineartime.Rmd:
````diff
@@ -26,7 +26,7 @@ library(haven)
 ## Load Data
 ```{r}
-stan_data <- readRDS(here("1_IGT_PP", "Data", "1_Preprocessed", "stan_ready_ORL_joint_retest.rds"))
+stan_data <- readRDS(here("1_IGT_PP", "Data", "1_Preprocessed", "stan_ready_ORL_IGT.rds"))
````
I think this name here should be `stan_ready_IGT_ORL.rds` instead of this? e.g.,: 

```suggestion
stan_data <- readRDS(here("1_IGT_PP", "Data", "1_Preprocessed", "stan_ready_IGT_ORL.rds"))
```

Nathaniel-Haines commented on 1_IGT_PP/Code/R/1_Analyses/fitting_joint_ORL_binomial_indMu.Rmd:
````diff
@@ -0,0 +1,284 @@
+
+## Load Data
+```{r}
+igt_data = readRDS(here("1_IGT_PP", "Data", "1_Preprocessed", "stan_ready_ORL_IGT.rds"))
````
Same here re the name needing to be `stan_ready_IGT_ORL.rds`

Nathaniel-Haines reviewed (changes requested):
@jeremy-haynes this all looks good! Nice job putting together the joint self-report + ORL model, it checks out on my end 👍 

I just made a comment re the path name to the stan data needing changed (based on what the name is when preprocessed+saved out). Actually, I wonder if that had something to do with the trouble running the model on your end? e.g., perhaps you have data with the current name, but it is not the data we mean to pass to the model? Because after running the preprocessing code and then changing the name as I suggest in the comment, the model does sample without issue on my end 🤔 

The only other high-level comment I have is that, when it comes to these types of joint models, it is really important to check that the ordering of participants in the `stan_data` is the same across behavioral and self-report measures. Perhaps we could save the subject IDs in the stan list when doing the preprocessing for both the ORL and self-reports, and then check that the lists are identical when bringing the data together for the modeling step?

Nathaniel-Haines commented on 1_IGT_PP/Code/Stan/igt_selfreport_orl_binomial_joint_indMu.stan:
```diff
@@ -0,0 +1,339 @@
+    int M = cols(R);
+    matrix[N,M] Sigma = quad_form_diag(R, sigma);
+    matrix[n_samples, M] X_sim;
+    for (i in 1:n_samples) {
+      X_sim[i,] = Phi_approx(multi_normal_rng(mu, Sigma)');
+    }
```
I realized that this will actually need changed now—the `Phi_approx` part is only necessary for parameters that use that transformation in the `transformed parameters` block. This solution is a bit ugly, but will transform the right parameters (`Arew`, `Apun`, `K`, and `theta`) to then estimate the correlation on the transformed scale the right way: 
```suggestion
    for (i in 1:n_samples) {
      X_sim[i,] = multi_normal_rng(mu, Sigma)';
    }
    // apply transformation as needed
    X_sim[,1] = Phi_approx(X_sim[,1]);
    X_sim[,2] = Phi_approx(X_sim[,2]);
    X_sim[,3] = Phi_approx(X_sim[,3]);
    X_sim[,6] = Phi_approx(X_sim[,6]);
```

jeremy-haynes commented:
@Nathaniel-Haines that sounds good! The incorrect data-referencing would be an issue haha! This all sounds good; I'll get to work on updating some of this. And I definitely agree, adding the subject IDs is wise to make sure it's all referencing the right people. I'll get to work on that as well. Thanks for checking out!

jeremy-haynes commented:
Hey @Nathaniel-Haines ,

I wanted to touch base with you on two things with fitting the Joint ORL-Binomial models for the IGT & self-report data (before I forget them😅).

As an update, so far, the models seem to be fitting fine. I suppose the problem I had before with the "rejecting initial value..." did have something to do with referencing the wrong datafile 🤷‍♂️. I've been fitting the models one at a time and they take a while but so far so good.

The first thing I wanted to touch base with you on is how I added checks for the IDs between the IGT and self-report data. I added code to the preprocessing files to save the participant IDs for the [IGT data](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/0_Preprocess/IGT_Preprocess.R#L382-L418) and the [self-report data](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/0_Preprocess/SelfReport_Preprocess.R#L118-L158), and then in the RMD for fitting the model, I put in a check (see [here](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/1_Analyses/fitting_joint_ORL_binomial_indMu.Rmd#L43-L57)) to make sure that those are the same. (btw, the permalinks are super cool on here!). I feel good that those are working correctly and that the IDs are the same across the IGT and self-report data, but if you notice anything, let me know.

The second thing I wanted to check with you about is making sure I'm labeling the posteriors correctly for the correlation matrix. Based on how the correlation matrix is set up and how we use the SD derived from that correlation matrix in the [joint ORL-binomial stan code](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/Stan/igt_selfreport_orl_binomial_joint_indMu.stan#L100-L111), I believe that we are essentially setting up the following 6x6 correlation matrix:
<img width="491" alt="image" src="https://github.com/hollysully/IGT_PP_Shared/assets/126425405/4af3e3c0-5ea9-41c0-9402-86fe657205f9">

If we extract the posterior distributions from the stan-fit, we get a 16000x6x6 array of correlations coefficients (i.e., `orl_posteriors <- extract(orl_fit)... orl_posteriors$R`). Rows should be iterations and I'm assuming the columns and matrix slices correspond to the rows and columns, respectively, of the correlation matrix above. Would that be the correct assumption? For example, `orl_posteriors$R[,1,1]` should be the posterior distribution of Rs between Arew and Arew (thus, all equal to 1 which it is), `orl_posteriors$R[,2,1]` should be the posterior distribution of Rs between Arew and Apun... I have some, admittedly complicated, code that runs through the posterior distribution of Rs in this way ([here](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/1_Analyses/fitting_joint_ORL_binomial_indMu.Rmd#L172-L201)). Let me know what you think.

Thanks,
Jeremy

Nathaniel-Haines commented on 1_IGT_PP/Code/R/0_Preprocess/IGT_Preprocess.R:
```diff
@@ -392,11 +394,14 @@ for (i in 1:numSubjs) {
-
+# Check that how the for loop above loops through the subject IDs is in the same order as subjList
+(condensedIDs = apply(IDs, c(1,3), mean)) # Visually check they are the same (-1 means missing data on that session)
+ifelse(mean(condensedIDs[,1] == subjList) == 1, "Equal", "Not Equal") # Check if IDs in for loop equal IDs in subjList
```
Instead of this, you can just explicitly check that they are all identical using the following syntax (makes the intention a little more clear IMO): 

```suggestion
ifelse(all(stan_data$igt_IDs == stan_data$sr_IDs), "Equal", "Not Equal") # Check if IDs in for loop equal IDs in subjList
```

Nathaniel-Haines commented:
> The first thing I wanted to touch base with you on is how I added checks for the IDs between the IGT and self-report data. I added code to the preprocessing files to save the participant IDs for the [IGT data](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/0_Preprocess/IGT_Preprocess.R#L382-L418) and the [self-report data](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/0_Preprocess/SelfReport_Preprocess.R#L118-L158), and then in the RMD for fitting the model, I put in a check (see [here](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/1_Analyses/fitting_joint_ORL_binomial_indMu.Rmd#L43-L57)) to make sure that those are the same. (btw, the permalinks are super cool on here!). I feel good that those are working correctly and that the IDs are the same across the IGT and self-report data, but if you notice anything, let me know.

@jeremy-haynes yes this check is great! Helps us know that everything is aligned properly. I made a minor suggestion about a way to check more explicitly, but definitely not necessary and more just a readability/style thing. 

> If we extract the posterior distributions from the stan-fit, we get a 16000x6x6 array of correlations coefficients (i.e., orl_posteriors <- extract(orl_fit)... orl_posteriors$R). Rows should be iterations and I'm assuming the columns and matrix slices correspond to the rows and columns, respectively, of the correlation matrix above. Would that be the correct assumption? For example, orl_posteriors$R[,1,1] should be the posterior distribution of Rs between Arew and Arew (thus, all equal to 1 which it is), orl_posteriors$R[,2,1] should be the posterior distribution of Rs between Arew and Apun... I have some, admittedly complicated, code that runs through the posterior distribution of Rs in this way ([here](https://github.com/hollysully/IGT_PP_Shared/blob/e844397e1acb65354d274f0cb3f3c4c1fd53b146/1_IGT_PP/Code/R/1_Analyses/fitting_joint_ORL_binomial_indMu.Rmd#L172-L201)). Let me know what you think.

Yes you are 100% correct! Basically, each iteration, `R[i,1,2]` is a correlation matrix for the given indices (1 and 2 in this case). So, the posterior on the correlation is the distribution of values for the given indices across all iterations `i`. If you want to easily just get the mean correlation matrix, you can do something like `apply(R, c(2,3), mean)`, and it will return the mean for each posterior correlation in the format of the matrix itself. More generally, the `apply` function is great for summarizing posteriors that are in the form of a matrix. You just specify which dimensions you want to get the result back in (i.e. passing in `c(2,3)` means I want to get back the second and third indices), and then pass in any function of interest (e.g., `mean`, `sd`), and you will get a result back that is a bit easier to interpret given it retains the matrix format.

jeremy-haynes commented:
@Nathaniel-Haines thanks for checking these out! I definitely support more readability, so committed that. And that sounds good about the correlation matrix. That's awesome I could use apply. I definitely made that way harder than it needed to be. Thanks!

jeremy-haynes commented:
Hey @Nathaniel-Haines , I've run into an issue with two of the joint ORL-binomial models that I'm a little stuck on. Specifically, the models fit to basfunsk and BIS self-report data return the warning/error message below (I only put basfunsk plots for illustration).

![image](https://github.com/hollysully/IGT_PP_Shared/assets/126425405/5c180f8d-55f6-4e21-ae37-7f073971b8bd)

It seems to have issues with mus and choice_pred values, but I don't see the issue in the output. For example, when I make traceplots of the mus, I get the following

<img width="499" alt="image" src="https://github.com/hollysully/IGT_PP_Shared/assets/126425405/d289299e-36d1-43e2-821e-6b66e18b7fc4">

I've looked at the self-report data and I'm not seeing an issue. In addition, I would think that that would not affect the ORL parameters (although I assume it all interacts to some extent). Do you have any thoughts?

Nathaniel-Haines commented:
> Hey @Nathaniel-Haines , I've run into an issue with two of the joint ORL-binomial models that I'm a little stuck on. Specifically, the models fit to basfunsk and BIS self-report data return the warning/error message below (I only put basfunsk plots for illustration).

This sort of thing will normally happen if the variable type is specified but never assigned. However, everything seems to be assigned as normal.. I am running some tests now to diagnose 🤔

Nathaniel-Haines reviewed (commented):
@jeremy-haynes so I have tried a few things, and I am a bit stumped as to why we are getting the strange warnings. Like I noted, it usually happens if variables are left unspecified, but in our case I cannot seem to find anywhere that this would be happening. Also, the outputs are all normal, as you mention... I will take a deeper dive by end of this next weekend. In the meantime, it may be worth exploring which datasets this happens to. If it is just this one, then it could have something to do with the data 🤔 

EDIT: I tried fitting the model on the `bastot` scale (which has two sessions of data akin to the `basfunsk` scale) and I did not get any warnings, so I think it has something to do with the `basfunsk` data we are passing into stan. Not sure exactly what it could be, but seems to be the culprit! Everything re the stan code itself checks out OK to me 👍

jeremy-haynes commented on 1_IGT_PP/Code/Stan/igt_selfreport_orl_binomial_joint_indMu.stan:
```diff
@@ -0,0 +1,344 @@
+        
+        for (t in 1:Tsubj[i,s]) { // Run through RL algorithm trial-by-trial
+          
+        // ORL LIKELIHOOD - predict choice as a function of utility
+        choice[i,t,s] ~ categorical_logit(to_vector(utility[card[i,t,s], :]));
+        
+        // After choice, calculate prediction error
+        PEval      = outcome[i,t,s] - ev[card[i,t,s], choice[i,t,s]];     // Value prediction error
+        PEfreq     = sign[i,t,s] - ef[card[i,t,s], choice[i,t,s]];        // Win-frequency prediction error
+        PEfreq_fic = -sign[i,t,s]/1 - ef[card[i,t,s], (3-choice[i,t,s])]; // Lose-frequency prediction error?
+        
+        if (outcome[i,t,s] >= 0) {  // If participant DID NOT lose
+          // Update expected win-frequency of what participant DID NOT chose to do
+          ef[card[i,t,s], (3-choice[i,t,s])] = ef[card[i,t,s], (3-choice[i,t,s])] + Apun[i] * PEfreq_fic;
+          // Update what participant chose
+          ef[card[i,t,s], choice[i,t,s]] = ef[card[i,t,s], choice[i,t,s]] + Arew[i] * PEfreq;
+          ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Arew[i] * PEval;
+        } else { // If participant DID lose
+          // Update expected win-frequency of what participant DID NOT choose to do
+          ef[card[i,t,s], (3-choice[i,t,s])] = ef[card[i,t,s], (3-choice[i,t,s])] + Arew[i] * PEfreq_fic;
+          // Update what participant chose
+          ef[card[i,t,s], choice[i,t,s]] = ef[card[i,t,s], choice[i,t,s]] + Apun[i] * PEfreq;
+          ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Apun[i] * PEval;
+        }
+        
+        // Update perseverance
+        pers[card[i,t,s], choice[i,t,s]] = 1;
+        pers = pers / (1 + K_tr);
+        
+        // Calculate expected value of card
+        utility = ev + ef * betaF[i] + pers * betaP[i];
```
Ahh good catch, I went ahead and fixed it and the whitespace issue below in a recent commit.

jeremy-haynes replied:
Oh and just realized that you fixed the whitespace one!

jeremy-haynes commented:
> @jeremy-haynes so I have tried a few things, and I am a bit stumped as to why we are getting the strange warnings. Like I noted, it usually happens if variables are left unspecified, but in our case I cannot seem to find anywhere that this would be happening. Also, the outputs are all normal, as you mention... I will take a deeper dive by end of this next weekend. In the meantime, it may be worth exploring which datasets this happens to. If it is just this one, then it could have something to do with the data 🤔
> 
> EDIT: I tried fitting the model on the `bastot` scale (which has two sessions of data akin to the `basfunsk` scale) and I did not get any warnings, so I think it has something to do with the `basfunsk` data we are passing into stan. Not sure exactly what it could be, but seems to be the culprit! Everything re the stan code itself checks out OK to me 👍

Ok, I'll check out the data some more. It also seems to happen with the BIS data. Maybe I can compare those datasets to some of the others and see if there's something different that might be causing the issue. Anyway, thanks for checking this out! I'll let you know if I come up with anything.

Nathaniel-Haines commented:
> @Nathaniel-Haines Ok, I'll check out the data some more. It also seems to happen with the BIS data. Maybe I can compare those datasets to some of the others and see if there's something different that might be causing the issue. Anyway, thanks for checking this out! I'll let you know if I come up with anything.

@jeremy-haynes I think I figured it out! In the likelihood for the binomial model, we had it written like so (not exactly, but writing this way for demonstration): 

```
transformed parameters {
    for(i in 1:N) {  // Loop over participants
        ...
        theta[i] = Phi_approx(mu_p[6] + parameter_tilde[6,i]);
    }
}
...
model {
...
    for (i in 1:N) {         // Loop through individual participants
        for (s in 1:SR_sess) {
          if (missing[i,s] == 0) {
            score[i,s] ~ binomial(M, theta[i]);
          }
        }
...
}
...
```

After digging in a bit more, I thought that perhaps the issue could be with numerical instability in doing the `theta = Phi_approx(mu_p[6] + parameter_tilde[6,i])` and then `score ~ binomial(M, theta)` step. Basically, if the `mu_p[6] + parameter_tilde[6,i]` value is either really small or really large, transforming it and then putting the transformed value through the likelihood can lead to numerical underflow or overflow. e.g., try doing `pnorm(-40, 0, 1)` in R and see what happens. You get `0` despite the CDF of the normal distribution technically not being 0 (it's just that -40 is so far out in the tails that the CDF is so small that, numerically, the computer just cannot represent it). 

That said, this is typically dealt with by using functions that protect against overflow/underflow. e.g., in Stan, the functions like `categorical_logit(par_raw)` and `bernoulli_logit(par_raw)` take the place of `categorical(softmax(par_raw))` and `bernoulli(inv_logit(par_raw))`, but they deal with numerical instability better. In our code above, you see that we do the `Phi_approx` transform first, and then just use `bernoulli` in the likelihood. 

Unfortunately, there is not a `bernoulli_probit` function in Stan, so we cannot just replace the likelihood by that. Instead, I thought it would be best to use the `logit/inv_logit` transformation for `theta` as opposed to `probit/inv_probit` as it is set up now. I also modified some of the priors to be a bit tighter on `sigma`, resulting in the following Stan code: 

```
functions {
  // to compute empirical correlation matrix given input matrix
  matrix empirical_correlation(matrix X) {
    int N = rows(X);
    int M = cols(X);
    matrix[N, M] centered;
    vector[M] std_devs;
    matrix[M, M] Sigma;
    matrix[M, M] R;

    // calculate covariance matrix
    for (j in 1:M) {
      centered[,j] = X[,j] - mean(X[,j]);
    }
    Sigma = (centered' * centered) / (N - 1);

    // Calculate correlation matrix
    std_devs = sqrt(diagonal(Sigma));
    for (i in 1:M) {
      for (j in 1:M) {
        R[i,j] = Sigma[i,j] / (std_devs[i] * std_devs[j]);
      }
    }

    return R;
  }

  real Phi_approx_group_mean_rng(real mu, real sigma, int n_samples) {
    vector[n_samples] par;
    for (i in 1:n_samples) {
      par[i] = Phi_approx(normal_rng(mu, sigma));
    }
    return mean(par);
  }
  
  real inv_logit_group_mean_rng(real mu, real sigma, int n_samples) {
    vector[n_samples] par;
    for (i in 1:n_samples) {
      par[i] = inv_logit(normal_rng(mu, sigma));
    }
    return mean(par);
  }

  matrix Phi_approx_corr_rng(vector mu, vector sigma, matrix R, int n_samples) {
    int N = rows(R);
    int M = cols(R);
    matrix[N,M] Sigma = quad_form_diag(R, sigma);
    matrix[n_samples, M] X_sim;
    for (i in 1:n_samples) {
      X_sim[i,] = multi_normal_rng(mu, Sigma)';
    }
    // apply transformation as needed
    X_sim[,1] = Phi_approx(X_sim[,1]);
    X_sim[,2] = Phi_approx(X_sim[,2]);
    X_sim[,3] = Phi_approx(X_sim[,3]);
    X_sim[,6] = inv_logit(X_sim[,6]);
    return empirical_correlation(X_sim);
  }
}


data {
  int<lower=1> N; // Number of participants
  
  // IGT-SPECIFIC DATA
  int<lower=1> S;      // Number of sessions
  int<lower=1> T;      // Total possile number of trials
  int card[N,T,S];     // Cards presented on each trial
  int Tsubj[N,S];      // Total number of trials presented to each subject on each session
  int choice[N,T,S];   // Choices on each trial
  real outcome[N,T,S]; // Outcomes received on each trial
  real sign[N,T,S];    // Signs of the outcome received on each trial
  
  // SELF-REPORT-SPECIFIC DATA
  int SR_sess;            // Number of session's worth of self-report data
  int missing[N,SR_sess]; // Matrix of missingness for current scale
  int score[N,SR_sess];   // Matrix of scores for current scale
  int M;                  // Max score for current scale
}


parameters {
  // Hyper(group)-parameters
  vector[6] mu_p;           // Vector of 5 ORL mus and 1 binomial Mu
  vector<lower=0>[6] sigma; // Vector of 5 ORL sigmas and 1 binomial sigma
  
  // Untransformed & non-correlated subject-level parameters
  matrix[6,N] parameter_pr;
  
  // Correlation matrices for correlating between parameters
  cholesky_factor_corr[6] R_chol;
}


transformed parameters {
  // Transformed & correlated subject-level parameters
  vector<lower=0,upper=1>[N] Arew;
  vector<lower=0,upper=1>[N] Apun;
  vector<lower=0,upper=5>[N] K;
  vector[N] betaF;
  vector[N] betaP;
  vector[N] theta;
  
  // Untransformed but correlated subject-level parameters
  matrix[6,N] parameter_tilde;
  
  // Untransformed subject-level parameters incorporating correlation between parameters
  parameter_tilde = diag_pre_multiply(sigma, R_chol) * parameter_pr;
  
  // Transform parameters to use in likelihoods
  for(i in 1:N){  // Loop over participants
    Arew[i]  = Phi_approx(mu_p[1] + parameter_tilde[1,i]);
    Apun[i]  = Phi_approx(mu_p[2] + parameter_tilde[2,i]);
    K[i]     = Phi_approx(mu_p[3] + parameter_tilde[3,i]) * 5;
    betaF[i] = mu_p[4] + parameter_tilde[4,i];
    betaP[i] = mu_p[5] + parameter_tilde[5,i];
    theta[i] = mu_p[6] + parameter_tilde[6,i];
  }
}


model {
  // Declare variables to calculate utility after each trial for ORL: These 4 (number of cards) x 2 (playing vs. not playing) matrices
  matrix[4,2] ef;
  matrix[4,2] ev;
  matrix[4,2] pers;
  matrix[4,2] utility;
    
  real PEval;
  real PEfreq;
  real PEfreq_fic;
  real K_tr;
  
  // Hyperpriors
  mu_p  ~ normal(0, 1);
  // sigma ~ normal(0, 1);      // Arew
  sigma[1:3] ~ normal(0, 0.2); // Apun
  sigma[4:5] ~ normal(0, 1.0); // Apun
  sigma[6] ~ normal(0, 0.5); // theta
  
  R_chol ~ lkj_corr_cholesky(1); // Correlation matrix
  
  // Priors for untransformed and non-correated subject-level parameters
  to_vector(parameter_pr) ~ normal(0, 1.0);
  
  // Loop through both sets of data
  for (i in 1:N) {         // Loop through individual participants
    for (s in 1:SR_sess) {
      if (missing[i,s] == 0) {
        score[i,s] ~ binomial_logit(M, theta[i]);
      }
    }
    
    for (s in 1:S) {       // Loop though sessions for participant i
    
      // SELF-REPORT LIKELIHOOD - predict number of "endorsements" out of max
      if (Tsubj[i,s] > 0) {    // If we have data for participant i on session s, run through RL algorithm
        
        // Initialize starting values
        K_tr = pow(3, K[i]) - 1;
        for (pp in 1:2) {  // Looping over play/pass columns and assigning each 0 to each card
          ev[:,pp] = rep_vector(0,4);
          ef[:,pp] = rep_vector(0,4);
          pers[:,pp] = rep_vector(0,4);
          utility[:,pp] = rep_vector(0,4);
        }
        
        for (t in 1:Tsubj[i,s]) { // Run through RL algorithm trial-by-trial
          
          // ORL LIKELIHOOD - predict choice as a function of utility
          choice[i,t,s] ~ categorical_logit(to_vector(utility[card[i,t,s], :]));
          
          // After choice, calculate prediction error
          PEval      = outcome[i,t,s] - ev[card[i,t,s], choice[i,t,s]];     // Value prediction error
          PEfreq     = sign[i,t,s] - ef[card[i,t,s], choice[i,t,s]];        // Win-frequency prediction error
          PEfreq_fic = -sign[i,t,s]/1 - ef[card[i,t,s], (3-choice[i,t,s])]; // Lose-frequency prediction error?
          
          if (outcome[i,t,s] >= 0) {  // If participant DID NOT lose
            // Update expected win-frequency of what participant DID NOT chose to do
            ef[card[i,t,s], (3-choice[i,t,s])] = ef[card[i,t,s], (3-choice[i,t,s])] + Apun[i] * PEfreq_fic;
            // Update what participant chose
            ef[card[i,t,s], choice[i,t,s]] = ef[card[i,t,s], choice[i,t,s]] + Arew[i] * PEfreq;
            ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Arew[i] * PEval;
          } else { // If participant DID lose
            // Update expected win-frequency of what participant DID NOT choose to do
            ef[card[i,t,s], (3-choice[i,t,s])] = ef[card[i,t,s], (3-choice[i,t,s])] + Arew[i] * PEfreq_fic;
            // Update what participant chose
            ef[card[i,t,s], choice[i,t,s]] = ef[card[i,t,s], choice[i,t,s]] + Apun[i] * PEfreq;
            ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Apun[i] * PEval;
          }
          
          // Update perseverance
          pers[card[i,t,s], choice[i,t,s]] = 1;
          pers = pers / (1 + K_tr);
          
          // Calculate expected value of card
          utility = ev + ef * betaF[i] + pers * betaP[i];
        }
      }
    }
  }
}

generated quantities {
  // Hyper(group)-parameters 
  real<lower=0,upper=1> mu_Arew;
  real<lower=0,upper=1> mu_Apun;
  real<lower=0,upper=5> mu_K;
  real mu_betaF;
  real mu_betaP;
  real mu_theta;

  // For posterior predictive check
  real choice_pred[N,T,S];
  real score_pred[N,S];

  // test-retest correlations
  corr_matrix[6] R;
  
  // Reconstruct correlation matrix from cholesky factor
  R = Phi_approx_corr_rng(mu_p, sigma, R_chol * R_chol', 10000);
  
  // Set all posterior predictions to -1 (avoids NULL values)
  score_pred = rep_array(-1, N, SR_sess);
  choice_pred = rep_array(-1, N, T, S);
  
  // Compute group-level means
  mu_Arew  = Phi_approx_group_mean_rng(mu_p[1], sigma[1], 10000);
  mu_Apun  = Phi_approx_group_mean_rng(mu_p[2], sigma[2], 10000);
  mu_K     = Phi_approx_group_mean_rng(mu_p[3], sigma[3], 10000) * 5; 
  mu_betaF = mu_p[4];
  mu_betaP = mu_p[5];
  mu_theta = inv_logit_group_mean_rng(mu_p[6], sigma[6], 10000);
  
  { // local section, this saves time and space
    // Declare variables to calculate utility after each trial: These 4 (number of cards) x 2 (playing vs. not playing) matrices
    matrix[4,2] ef;
    matrix[4,2] ev;
    matrix[4,2] pers;
    matrix[4,2] utility;
      
    real PEval;
    real PEfreq;
    real PEfreq_fic;
    real K_tr;
  
    for (i in 1:N) {    // Loop through individual participants
      for (s in 1:SR_sess) {
        score_pred[i,s] = binomial_rng(M, inv_logit(theta[i]));
      }
      for (s in 1:S) {    // Loop though sessions for participant i
        if (Tsubj[i,s] > 0) {    // If we have data for participant i on session s, run through RL algorithm
          
          // Initialize starting values
          K_tr = pow(3, K[i]) - 1;
          for (pp in 1:2) {  // Looping over play/pass columns and assigning each 0 to each card
            ev[:,pp] = rep_vector(0,4);
            ef[:,pp] = rep_vector(0,4);
            pers[:,pp] = rep_vector(0,4);
            utility[:,pp] = rep_vector(0,4);
          }
          
          for (t in 1:Tsubj[i,s]) { // Run through RL algorithm trial-by-trial
            // ORL LIKELIHOOD - predict choice as a function of utility
            choice_pred[i,t,s] = categorical_rng(softmax(to_vector(utility[card[i,t,s], :])));
            
            // After choice, calculate prediction error
            PEval      = outcome[i,t,s] - ev[card[i,t,s], choice[i,t,s]];     // Value prediction error
            PEfreq     = sign[i,t,s] - ef[card[i,t,s], choice[i,t,s]];        // Win-frequency prediction error
            PEfreq_fic = -sign[i,t,s]/1 - ef[card[i,t,s], (3-choice[i,t,s])]; // Lose-frequency prediction error?
            
            if (outcome[i,t,s] >= 0) {  // If participant DID NOT lose
              // Update expected win-frequency of what participant DID NOT chose to do
              ef[card[i,t,s], (3-choice[i,t,s])] = ef[card[i,t,s], (3-choice[i,t,s])] + Apun[i] * PEfreq_fic;
              // Update what participant chose
              ef[card[i,t,s], choice[i,t,s]] = ef[card[i,t,s], choice[i,t,s]] + Arew[i] * PEfreq;
              ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Arew[i] * PEval;
            } else { // If participant DID lose
              // Update expected win-frequency of what participant DID NOT choose to do
              ef[card[i,t,s], (3-choice[i,t,s])] = ef[card[i,t,s], (3-choice[i,t,s])] + Arew[i] * PEfreq_fic;
              // Update what participant chose
              ef[card[i,t,s], choice[i,t,s]] = ef[card[i,t,s], choice[i,t,s]] + Apun[i] * PEfreq;
              ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Apun[i] * PEval;
            }
            
            // Update perseverance
            pers[card[i,t,s], choice[i,t,s]] = 1;
            pers = pers / (1 + K_tr);
            
            // Calculate expected value of card
            utility = ev + ef * betaF[i] + pers * betaP[i];
          }
        }
      }
    }
  }
}

```

Now, `theta` is on the untransformed scale, and we use `bernoulli_logit(M, theta)` in the likelihood instead of what we had before. Note that the generated quantities block still does the transform in two steps because there are no `*_logit_rng` functions. I also had to use `inv_logit_*` functions in place of `Phi_approx_*` functions for `theta` as needed. 

With this model, I do not get the warnings like we did before when fitting to the `basfunsk` scale! Let me know if you have any questions, I know this is a bit in the weeds ha. But this type of numerical instability tends to come up frequently when working with bounded parameters, so it is good to know about!

Nathaniel-Haines commented:
@jeremy-haynes OK after checking more, I guess it was not actually instability per se as outlined above—it seems it was actually my tightening up the priors on `sigma` that led to no warnings. e.g., before you had: 

```
sigma ~ normal(0,1);
```

but I changed to: 

```
sigma[1:3] ~ normal(0, 0.2);
sigma[4:5] ~ normal(0, 1.0);
sigma[6] ~ normal(0, 0.2); 
```

I made this change because `N(0,1)` tends to allow for quite a wide prior for the bounded parameters. e.g., if you simulate out:

```
# parameter group mean+sd
mu = rnorm(1000, 0, 1)
sigma = truncnorm::rtruncnorm(1000, a=0, mean=0, sd=1)

# implied prior on person-level par
par = rnorm(1000, mu, sigma)
hist(pnorm(par))
```

You will see that the prior puts a lot of mass toward the edges of the parameter space. Tightening to `N(0,1)` gives something I think is more uniform: 

```
# parameter group mean+sd
mu = rnorm(1000, 0, 1)
sigma = truncnorm::rtruncnorm(1000, a=0, mean=0, sd=.2)

# implied prior on person-level par
par = rnorm(1000, mu, sigma)
hist(pnorm(par))
```

Anyway, when I use the tighter priors, warnings disappear! Was there a reason you loosened the prior on `sigma` before?? 

(TLDR: it was the priors on `sigma` this whole time, not numerical stability as I originally thought)

jeremy-haynes commented:
@Nathaniel-Haines ahh shoot, that was totally my bad. I changed the sigmas to that because for some reason I had it in my head that they should all be the same for the correlations, but I suppose it probably doesn't matter because they're all relative anyway, right? Sorry about that. That does make sense though on why that happened.

jeremy-haynes commented:
Hey @Nathaniel-Haines ,

I added the carryover parameter to the joint model. I did a quick fit (2000 iterations w/1000 warmup) and everything seems relatively similar to as before. It might be worth double-checking that I implemented it correctly. The RMD for running it is [here](1_IGT_PP/Code/R/1_Analyses/fitting_ORL_carryover_joint.Rmd). I added several lines of code to the joint model in a new stan file [here](1_IGT_PP/Code/Stan/igt_orl_carryover_joint.stan), but the area that may need a careful look at it [here](https://github.com/hollysully/IGT_PP_Shared/blob/d88df63c93625e518ab57beafa07764b4cc48105/1_IGT_PP/Code/Stan/igt_orl_carryover_joint.stan#L180-L196). Let me know what you think!

Jeremy

jeremy-haynes commented:
Darn, my last comment's hyperlinks are not seeming to work. Essentially, I added the files in the commit labeled "added carryover parameter to joint model (cow)" and the important lines of stan code are 180-195.

Nathaniel-Haines commented:
> I added the carryover parameter to the joint model. I did a quick fit (2000 iterations w/1000 warmup) and everything seems relatively similar to as before. It might be worth double-checking that I implemented it correctly. The RMD for running it is [here](1_IGT_PP/Code/R/1_Analyses/fitting_ORL_carryover_joint.Rmd). I added several lines of code to the joint model in a new stan file [here](1_IGT_PP/Code/Stan/igt_orl_carryover_joint.stan), but the area that may need a careful look at it [here](https://github.com/hollysully/IGT_PP_Shared/blob/d88df63c93625e518ab57beafa07764b4cc48105/1_IGT_PP/Code/Stan/igt_orl_carryover_joint.stan#L180-L196). Let me know what you think!

@jeremy-haynes I looked over and you implemented it well! The only issue I could note was that, you would want to add the weighting to the `generated quantities` block as well in order to generate the posterior predictions and log likelihood values correctly. But that would not influence the parameters estimates themselves. I ran a small model too and it seems that things are like you said—not a huge difference here (although the weighting parameter is quite high! suggesting there are some carry-over effects, just not anything super substantial). 

That said, the estimates themselves are just rather uncertain, they aren't necessarily low. e.g., here is an example of the correlation I get for `Arew`: 

<img width="833" alt="Screen Shot 2023-07-16 at 4 26 12 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/1052a5f0-05ad-4e14-8c02-fd9b00c8cc82">

The mean is low-ish, and the CI is indeed over 0, but it is primarily due to the fact that uncertainty is just so high. I assume that results would be a little more clear with more data, but for now I would not say that this necessarily bodes poorly for reliability. It could be that the play-pass version is just less informative than the original version wrt the model parameters 🤔 

I think that it would be reasonable to more forward with just the original test-retest model results—nobody has published results for this particular model-task combination, and it could be informative for others interested in the task. Curious what you think?

Nathaniel-Haines commented:
I have been playing around with different fictive updating rules, and one that I tried was the following: 

```
// Likelihood - predict choice as a function of utility
        choice[i,t,s] ~ categorical_logit(to_vector(utility[card[i,t,s], :]));
        
        // After choice, calculate prediction error
        PEval      = outcome[i,t,s] - ev[card[i,t,s], choice[i,t,s]];     // Value prediction error
        PEfreq     = sign[i,t,s] - ef[card[i,t,s], choice[i,t,s]];        // Win-frequency prediction error
        PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]]; // Lose-frequency prediction error?
        ef_chosen = ef[card[i,t,s], choice[i,t,s]];
        
        if (outcome[i,t,s] >= 0) {  // If participant DID NOT lose
          // Update expected win-frequency of what participant DID NOT chose to do
          ef[:, choice[i,t,s]] = ef[:, choice[i,t,s]] + Apun[i,s] * PEfreq_fic;
          // Update what participant chose
          ef[card[i,t,s], choice[i,t,s]] = ef_chosen + Arew[i,s] * PEfreq;
          ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Arew[i,s] * PEval;
        } else { // If participant DID lose
          // Update expected win-frequency of what participant DID NOT choose to do
          ef[:, choice[i,t,s]] = ef[:, choice[i,t,s]] + Arew[i,s] * PEfreq_fic;
          // Update what participant chose
          ef[card[i,t,s], choice[i,t,s]] = ef_chosen + Apun[i,s] * PEfreq;
          ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Apun[i,s] * PEval;
        }
```

in the ORL for the original IGT, the fictive updating is across decks given that there is only one choice option per deck (i.e. just play, no pass). When generalizing the model, I original decided to make the fictive updating work within decks (i.e. between play and pass) as opposed to across decks. The code above does fictive updating across decks for the same option (i.e. if play was chosen, play is fictively updated across decks). 

When fitting this model, things look a bit different. Basically, reliability looks much better for `Arew`, slightly worse for `Apun`, and similar (if I remember correctly) for other parameters: 

<img width="839" alt="Screen Shot 2023-07-16 at 8 28 39 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/f35627b6-eb88-440e-8b09-7a54e30e971e">

Also, the posterior predictions look similar to before, indicating that fit is not compromised by this change: 

<img width="856" alt="Screen Shot 2023-07-16 at 8 24 19 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/fda2c99c-b017-4d56-b3f4-c3b0e8f21283">

Anyway, could be worth exploring this model more! Let me know what you think, and if I can answer any questions this all brings to mind. (NOTE: A couple chains were not converging well for the model I fit, in part likely due to the fact that I fit so few samples. We might need to modify priors if this continues with more samples, but would first want to just run the model with all samples to see if it is still an issue. The plots I have above are after removing those bad chains)

Nathaniel-Haines commented:
For completeness, here are posterior predictions for the session two data from the same model (above are session 1):

<img width="852" alt="Screen Shot 2023-07-16 at 8 40 34 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/d750e212-801f-46d5-9692-2b1dcc92726b">

hollysully commented:
Wow! Changing the fictive updating really changed reliability for Arew—that is exciting!

jeremy-haynes commented:
Hey @Nathaniel-Haines ,

I wanted to touch base on how the model-fitting is going based on our conversation last week. I was able to fit the updated fictive models to the IGT and SIGT data. One model uses the fictive updating as your coded wherein fictive updating occurs across decks for plays and passes with the updating being independent with respect to plays and passes. That is, we replaced `PEfreq_fic = -sign[i,t,s]/1 - ef[card[i,t,s], (3-choice[i,t,s])];` in the original joint model with `PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]];` in what I call the [updatedfic model](https://github.com/hollysully/IGT_PP_Shared/blob/5998d898bb1504e6af07e39d99a433c6d3743f40/1_IGT_PP/Code/Stan/igt_orl_joint_updatedfic.stan). The other model only updates fictive learning across decks for plays, not passes. Essentially, I did the following in the stan model I call [playfic model](https://github.com/hollysully/IGT_PP_Shared/blob/5998d898bb1504e6af07e39d99a433c6d3743f40/1_IGT_PP/Code/Stan/igt_orl_joint_playfic.stan):
`if(choice[i,t,s]==1) {
    PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]];
} else {
    PEfreq_fic = ef[:, choice[i,t,s]];
}`
When I fit these models to the data, two odd things happen.

First, when I fit the joint versions of both models to the 2-session IGT data, the chains look terrible for session 1 parameters; session 2 parameter chains look relatively good I think. Below are the group-level parameter traceplots from the playfic model (person-level parameter chains show a similar pattern):
<img width="601" alt="image" src="https://github.com/hollysully/IGT_PP_Shared/assets/126425405/f7030dc8-97eb-4d89-a21c-4a245a2d855e">
I am currently refitting one of the models with tighter priors on the mus... normal(0,.5), but to be honest, that's a bit of a guess.

The second oddity is that I believe that these two models are identical. The estimates and traceplots look the same for the joint models AND for the models I fit to the SIGT data (the monetary task). For the SIGT, those fit better; the traceplots look fine although there are 32 divergent transitions (for both types of fictive updating models). I'm working on PPC plots for the SIGT data now.

If you have any insights or thoughts, let me know. Also, in my last push, I caused a conflict (I think because I updated the regular ORL model from the SIGT). I'm not 100% sure the best way to resolve it: should I just delete the lines I think are problematic and mark it as resolved? Anyway, thanks and I hope all is well!

Nathaniel-Haines commented:
@jeremy-haynes Thanks for the update! Ah yeah it looks like convergence is still an issue with more iterations.. I was worried about this, as it showed up a bit with the small number of samples I used originally. I think the issue is not actually too bad though—if you do `traceplot(orl_fit, pars = c("mu_p", "lp__"))`, you will see that the log posterior is much lower for the bad chain than for the good chains. To me, this indicates that the bad chain is just sampling away in a non-optimal region of the posterior, and that we should be able to fix this with more informative priors, a different parameterization of the model, or even by setting better initial values for the chains. This all contrasts a case where, e.g., the log posterior for the "bad" chain is similar to the "good" chains, which would indicate more of a parameter identifiability issue re the structure of the model itself. 

In either case, I am working on the three different potential solutions I listed above and will report back once I made headway. re this point: 

> Essentially, I did the following in the stan model I call [playfic model](https://github.com/hollysully/IGT_PP_Shared/blob/5998d898bb1504e6af07e39d99a433c6d3743f40/1_IGT_PP/Code/Stan/igt_orl_joint_playfic.stan):
if(choice[i,t,s]==1) { PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]]; } else { PEfreq_fic = ef[:, choice[i,t,s]]; }

I think this is likely not the best implementation for only having fictive updating for play. The above is still doing fictive updating for pass, but it is setting the fictive prediction error to the current EF value. If we simply want to not update the EF values for unchosen decks when passing, the following rule would achieve that: 

```
if(choice[i,t,s]==1) { 
    PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]]; 
} else { 
    PEfreq_fic = rep_vector(0, 4); 
}
``` 

Then, in the updating equation, the fictive update will have no effect on the EF since the fictive PE is 0 for all decks. Let me know if that makes sense!

Nathaniel-Haines commented:
Oh and for the conflict—this is normal! It happened here because the `1_SIGT_PP/Code/Stan/igt_orl.stan` file on the main branch must have been changed, and you have also changed it here in this branch. Before merging this branch, we will just have to resolve the conflict. Happy to meet and walk through that process at some point!

Nathaniel-Haines reviewed (commented):
OK so after troubleshooting, I found that the convergence issues were a combination of the uninformative priors and the initial values used to kick-off MCMC sampling. 

Of these, it was mostly the initial values causing issues. By default, Stan will randomly sample between -2 and 2 to set initial values for each parameter for each chain. But sometimes a value in [-2,2] is not realistic given the model. Instead, you can pass an `init` argument to Stan that generates initial values for parameters of interest in a custom way. In the suggestion below, I am setting initial values for the group mean parameters (`mu_p`) in such a way that the initial values are randomly sampled from a region a bit closer to where we expect chains will end up. This basically fixes things! As seen below: 

![image](https://github.com/hollysully/IGT_PP_Shared/assets/25613611/3825df11-451d-4f20-9cb6-4cfa1069dd16)

However, I realized that more informative priors could also make estimation a bit more robust, (see the recommend the change below). Before, the `mu_p ~ normal(0, 1)` prior implied uniform priors on the group-level mean learning rates and K parameter. Learning rates are generally around .1 to .5 for models/tasks like the IGT, and so I set some priors to better capture this. When combined with the initial value function, this leads to not only better convergence, but K now shows some improvement in reliability from before:

![image](https://github.com/hollysully/IGT_PP_Shared/assets/25613611/ae4c007a-6852-4f3d-92ce-cb4484598e85)

Curious what you think! I have not applied this to the `playfic` model yet, but I imagine the same change to priors+initial values function should fix the issues there too 🤓 . Feel free to take or leave the prior recommendation—the initial value fix is the most important here.

jeremy-haynes commented:
> Hey @Nathaniel-Haines I just realized that I had implemented the model for the singe-timepoint ORL Playfic Playupdate Modified K. I'm re-running it now and will post the results of the model-comparison once they come in.

Ok, I just finished re-running the model and model-comparisons which are below.
                                  elpd_diff se_diff
ORL                                 0.0       0.0  
ORL Updatedfic                    -19.8      26.3  
ORL Playfic Playupdate Modified K -45.7      50.9

The original ORL is still the best-fitting although the standard errors of the differences are pretty high and would make the elpd_diff overlap with 0. From my understanding, this makes which model is the best-fitting a little ambiguous, but I'm less familiar with this type of model-comparison, so I'd be interested in your thoughts. For these, I only ran the model for 2000 iterations (with 1000 warmup) to get some results out before out meeting tomorrow, so I'm running it now as the larger model now which may finish before our meeting but I'm not 100% sure. Anyway, we can discuss stuff tomorrow. Hope all is well! - Jeremy

Nathaniel-Haines commented:
> Ok, I just finished re-running the model and model-comparisons which are below. elpd_diff se_diff ORL 0.0 0.0 ORL Updatedfic -19.8 26.3 ORL Playfic Playupdate Modified K -45.7 50.9
> 
> The original ORL is still the best-fitting although the standard errors of the differences are pretty high and would make the elpd_diff overlap with 0. From my understanding, this makes which model is the best-fitting a little ambiguous, but I'm less familiar with this type of model-comparison, so I'd be interested in your thoughts. For these, I only ran the model for 2000 iterations (with 1000 warmup) to get some results out before out meeting tomorrow, so I'm running it now as the larger model now which may finish before our meeting but I'm not 100% sure. Anyway, we can discuss stuff tomorrow. Hope all is well! - Jeremy

@jeremy-haynes yes you are exactly right! When the difference is this uncertain, it is safe to assume that we cannot make a clear judgement call that one model is better than the other. There are other things we could do to test the models with more scrutiny (e.g., fitting to just 1 timepoint, predicting for the same participants on the second timepoint and scoring models accordingly), but that will likely produce a similar conclusion. In cases like this, we can always report multiple models, but selecting a "final" model to use for inference comes down to more than just the fit. We can also report the estimates of interest from each model—it seems there is not a whole lot of variability between different models, so I don't expect the substantive results to vary my much either.

Nathaniel-Haines commented:
> As I was writing, I realized that one part of the playfic_playupdate_modifiedK model might need to be tweaked. Specifically, when a participant passes, perseverance is updated for plays across all decks. I wrote that as `pers[:,1] = pers[:,1] / (1 + K_tr);` (see also [here](https://github.com/hollysully/IGT_PP_Shared/blob/00e30b291cd533dd0d382fcc4952162a37ce75f6/1_IGT_PP/Code/Stan/igt_orl_joint_playfic_playupdate_modifiedK.stan#L205-L216)) but I think I misunderstood perseverance here. Shouldn't it still be
> 
> ```
> pers[card[i,t,s], 1] = 1;
> pers = pers / (1 + K_tr);
> ```
> 
> Let me know what you think

@jeremy-haynes apologies for the delayed responses! Things have been extra busy at Ledger, and I only now had some time to come back to this. 

Overall, I think this is a solid set of results! Your reasoning throughout the last few posts above makes sense to me, and I agree on converging toward the model that only updates on plays, and then does fictive updating across decks as opposed to within decks. I think this model is also true to the original ORL. 

re the perseverance rule, I think it should be the same result in either case. I think you can even simplify this code a bit by updating this part: 

```
if(choice[i,t,s]==1){
            // After choice, calculate prediction error
            PEval      = outcome[i,t,s] - ev[card[i,t,s], choice[i,t,s]]; 
            PEfreq     = sign[i,t,s] - ef[card[i,t,s], choice[i,t,s]];       
            PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]];
            ef_chosen  = ef[card[i,t,s], choice[i,t,s]];
            
            if (outcome[i,t,s] >= 0) {
              ef[:, choice[i,t,s]] = ef[:, choice[i,t,s]] + Apun[i,s] * PEfreq_fic;
              ef[card[i,t,s], choice[i,t,s]] = ef_chosen + Arew[i,s] * PEfreq;
              ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Arew[i,s] * PEval;
            } else { 
              ef[:, choice[i,t,s]] = ef[:, choice[i,t,s]] + Arew[i,s] * PEfreq_fic;
              ef[card[i,t,s], choice[i,t,s]] = ef_chosen + Apun[i,s] * PEfreq;
              ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Apun[i,s] * PEval;
            }
         
            pers[card[i,t,s], choice[i,t,s]] = 1;
            pers = pers / (1 + K_tr);
            
            utility = ev + ef * betaF[i,s] + pers * betaP[i,s];
          } else { // else pass
          pers[:,1] = pers[:,1] / (1 + K_tr);
          utility[:,1] = ev[:,1] + ef[:,1] * betaF[i,s] + pers[:,1] * betaP[i,s];
          }
        }
```

to: 

```
if(choice[i,t,s]==1){
            // After choice, calculate prediction error
            PEval      = outcome[i,t,s] - ev[card[i,t,s], choice[i,t,s]]; 
            PEfreq     = sign[i,t,s] - ef[card[i,t,s], choice[i,t,s]];       
            PEfreq_fic = -sign[i,t,s]/3 - ef[:, choice[i,t,s]];
            ef_chosen  = ef[card[i,t,s], choice[i,t,s]];
            
            if (outcome[i,t,s] >= 0) {
              ef[:, choice[i,t,s]] = ef[:, choice[i,t,s]] + Apun[i,s] * PEfreq_fic;
              ef[card[i,t,s], choice[i,t,s]] = ef_chosen + Arew[i,s] * PEfreq;
              ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Arew[i,s] * PEval;
            } else { 
              ef[:, choice[i,t,s]] = ef[:, choice[i,t,s]] + Arew[i,s] * PEfreq_fic;
              ef[card[i,t,s], choice[i,t,s]] = ef_chosen + Apun[i,s] * PEfreq;
              ev[card[i,t,s], choice[i,t,s]] = ev[card[i,t,s], choice[i,t,s]] + Apun[i,s] * PEval;
            }
         
            pers[card[i,t,s], choice[i,t,s]] = 1;
          }
        pers = pers / (1 + K_tr);
        utility = ev + ef * betaF[i,s] + pers * betaP[i,s];
        }
```

Reason being—if it is a play, updating occurs as normal per the `if` statement, and the `pers` term is set to 1 for the played deck. But then you can just get rid of the `else` statement entirely because you always want to decay `pers` regardless of play or pass (and the `pers` term for pass will always be 0 anyway). The `utility` is then always calculated at the end, no need to specify just updating the play index for each deck because the pass indices will be unchanged. 

Actually, now that I write this out, it is clear to me that `ev`, `ef`, and `pers` will now only take on non-zero values for the play options, and the pass option will always be 0. Therefore, `utility` will also only be non-0 for the play index. So there is now no need to have a play vs pass index for each deck. I will work on modifying the script to show what this change translates to in stan 🤓

Nathaniel-Haines commented:
@jeremy-haynes OK I have added the simplified version of the model [here](https://github.com/hollysully/IGT_PP_Shared/pull/11/commits/12327d7330970ebebe8223267a6a8a7d6a2d33af). Basically, I just got rid of some parts of the code that became unnecessary as we changed how it was functioning. I think this version is nice because it is easier to reason about despite it being identical to the `igt_orl_joint_playfic_playupdate_modifiedK.stan` version. I ran it with a few hundred iterations and am including some plots below to show that it is the same as what you had before: 

<img width="742" alt="Screen Shot 2023-08-20 at 2 45 51 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/8541186c-6443-4d24-af41-0cfc878048bc">

<img width="761" alt="Screen Shot 2023-08-20 at 2 46 03 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/8ed2b47c-f981-4d36-a7a7-5a88e0f110dd">

<img width="761" alt="Screen Shot 2023-08-20 at 2 47 39 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/009a8470-d9b1-4e7a-be70-3817f037edbe">

<img width="762" alt="Screen Shot 2023-08-20 at 2 49 32 PM" src="https://github.com/hollysully/IGT_PP_Shared/assets/25613611/8564974d-b28b-440a-a9cd-9972f32d5d22">

Anyway, I am signed off on moving forward with this model! Great job BTW in getting it to this point, I know from experience that it is quite time consuming (and often unclear IF you will get to a point where you are happy with it.. ha), but I would say all your work has paid off!

jeremy-haynes reviewed (commented):
Ahh yes, this is much simpler, cool!
